### PR TITLE
MD slicing algorithms should not add masked data to output workspace

### LINKED ISSUE #15781: MD Slicing algorithms should not add masked events to output workspace
SliceMD and BinMD should not add events to their output workspace if the events are in a masked box of the input workspace.

This is part of #14736.

### PR DESCRIPTION
**Pull request by matthew-d-jones**
Changes make SliceMD and BinMD not add events from masked boxes in the input workspace to the output workspace.

Release notes were updated (framework.rst Vates CLI section).

**To test:**

You can test using the following script.
`ws_masked_mdew` is a 3D workspace with one octant masked, so one octant should have NaN when viewed in the Slice Viewer. `sliced_ws` and `binned_ws` should have a signal of 0 in the corresponding region.

``` python
ws_mdew = CreateMDWorkspace(Dimensions='3', Extents='-2,2,-2,2,-2,2', Names='h,k,l',
                       Units='rlu,rlu,rlu', SplitInto='12')
FakeMDEventData(ws_mdew, UniformParams='1000', RandomizeSignal='1')
ws_mdew = SliceMD(ws_mdew,AlignedDim0='h,-2,2,12',AlignedDim1='k,-2,2,12',AlignedDim2='l,-2,2,12')
ws_masked_mdew = CloneMDWorkspace(ws_mdew)
MaskMD(Workspace=ws_masked_mdew,Dimensions="h,k,l",Extents="-2,0,-2,0,-2,0")

sliced_ws = SliceMD(InputWorkspace=ws_masked_mdew, AxisAligned=True, AlignedDim0='h,-2,2,12', AlignedDim1='k,-2,2,12', AlignedDim2='l,-2,2,12')
binned_ws = BinMD(InputWorkspace=ws_masked_mdew, AxisAligned=True, AlignedDim0='h,-2,2,12', AlignedDim1='k,-2,2,12', AlignedDim2='l,-2,2,12')
```

Fixes #15781.

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the coding standards? Is it well structured with small focussed classes/methods/functions?
- [x] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [x] If there are changes in the release notes then do they describe the changes appropriately?
##### Functional Tests
- [x] Do changes function as described? Add comments below that describe the tests performed?
- [x] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated? --> N/A
- [ ] Is user-facing documentation written in a user-friendly manner? --> N/A
- [ ] Has developer documentation been updated if required? --> N/A
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/MDAlgorithms/test/BinMDTest.h [hunk 1]
```diff
@@ -104,23 +104,17 @@ class BinMDTest : public CxxTest::TestSuite {
...
-                    std::string name2, std::string name3, std::string name4,
-                    double expected_signal, size_t expected_numBins,
-                    bool IterateEvents = true, size_t numEventsPerBox = 1,
-                    VMD expectBasisX = VMD(1, 0, 0),
+  void do_test_exec(const std::string &functionXML, const std::string &name1,
+                    const std::string &name2, const std::string &name3,
+                    const std::string &name4, const double expected_signal,
+                    const size_t expected_numBins, bool IterateEvents = true,
+                    size_t numEventsPerBox = 1, VMD expectBasisX = VMD(1, 0, 0),
                     VMD expectBasisY = VMD(0, 1, 0),
```
Comment (AntonPiccardoSelg) [inline Framework/MDAlgorithms/test/BinMDTest.h]: good

File: Framework/MDAlgorithms/test/BinMDTest.h [hunk 2]
```diff
@@ -128,6 +122,25 @@ class BinMDTest : public CxxTest::TestSuite {
...
+  execute_bin(const std::string &functionXML, const std::string &name1,
+              const std::string &name2, const std::string &name3,
+              const std::string &name4, const double expected_signal,
+              const size_t expected_numBins, bool IterateEvents,
+              size_t numEventsPerBox, VMD expectBasisX, VMD expectBasisY,
+              VMD expectBasisZ, IMDEventWorkspace_sptr in_ws) {
+    BinMD alg;
+    TS_ASSERT_THROWS_NOTHING(alg.initialize())
+    TS_ASSERT(alg.isInitialized())
+    Mantid::Kernel::SpecialCoordinateSystem appliedCoord =
```
Comment (AntonPiccardoSelg) [inline Framework/MDAlgorithms/test/BinMDTest.h]: could be `auto`

**Comment by AntonPiccardoSelg** [comment]
* Ran the test script provided above to generate the masked workspace which is then used by `SliceMD` and `BinMD`
* Inspecting the output in the VSI shows that the masked octant is not present any longer, ie `SliceMD` and `BinMD` respect masking

**Comment by FedeMPouzols** [comment]
Good! Note that when you have the release note update in a commit of its own as in this case (which is a very good idea) you can put a link to the commit in the PR description, like: https://github.com/mantidproject/mantid/pull/15787/commits/c2ab1417ab9ed5d8739dffac7d3d66ebbb0c6b02

**Pull request merged**